core/network-libp2p/src/custom_proto/handler.rs

```diff
@@ -347,14 +357,40 @@ where TMessage: CustomMessage, TSubstream: AsyncRead + AsyncWrite {
 				PerProtocolState::Poisoned
 			}
 
-			st @ PerProtocolState::Init(_) => {
+			PerProtocolState::Init { substreams, mut init_deadline } => {
+				match init_deadline.poll() {
+					Ok(Async::Ready(())) =>
+						error!(target: "sub-libp2p", "Handler initialization process is too long"),
+					Ok(Async::NotReady) => {}
+					Err(_) => error!(target: "sub-libp2p", "Tokio timer has errored")
+				}
+
 				return_value = None;
-				st
+				PerProtocolState::Init { substreams, init_deadline }
 			}
 
-			st @ PerProtocolState::Opening { .. } => {
-				return_value = None;
-				st
+			PerProtocolState::Opening { mut deadline } => {
+				match deadline.poll() {
+					Ok(Async::Ready(())) => {
+						deadline.reset(Instant::now() + Duration::from_secs(60));
```

Comment:
So we just reset it? Shouldn't we abort the process?

Reply:
Well, we report a misbehaviour and continue waiting.
I don't think there's any other good solution. If we switch to "disabled" mode, then we would have to add some logic to the misbehaviour event to indicate the outer layer that the handler gets disabled.
We could add a different state named `OpeningFailed` or something, but that would be overkill IMO.